### PR TITLE
C# 11: Generic attributes

### PR DESCRIPTION
**Pull request by michaelnebel**
It turns out that the extractor supports generic attributes out of the box due to the existing reference design between attributes and types.
That is, we only need to implement some minor Code QL library functionality for generic attributes to fetch type arguments and the like.
However, as a part of the process @tamasvajk discovered that attributes on methods in CIL was not correctly extracted. With the help of @tamasvajk this is fixed in https://github.com/github/codeql/pull/11907

### CODE REVIEW

File: csharp/ql/lib/semmle/code/csharp/Attribute.qll [hunk 1]
```diff
@@ -99,6 +105,31 @@ class Attribute extends TopLevelExprParent, @attribute {
...
+/**
+ * A generic attribute, for example `[...]` on line 1 in
+ *
+ * ```csharp
+ * [MyGenericAttribute<int>(0)]
+ * public void SomeMethod(string s) { }
+ * ```
+ */
+class GenericAttribute extends Attribute {
+  private ConstructedGeneric type;
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/Attribute.qll:117]: Maybe this should be ConstructedType.

Comment (tamasvajk) [reply csharp/ql/lib/semmle/code/csharp/Attribute.qll:117]: I think you're right. (It probably doesn't change much as there's only one type (`ConstructedType`) that's both `ConstructedGeneric` and `Type`, which is the return type of `Attribute::getType` used in the char pred)

File: csharp/ql/lib/change-notes/2023-01-03-genericattributes.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+* Added library support for generic attributes.
```
Comment (tamasvajk) [inline csharp/ql/lib/change-notes/2023-01-03-genericattributes.md:4]: I would also add that `cil.ConstructedType::getName` was changed. If anyone relied on that it's a breaking change, isn't it?

**Comment by michaelnebel** [comment]
> @michaelnebel It might be worth checking if the CIL extractor also supports generic attributes. We do some attribute processing in [`CustomAttributeDecoder`](https://github.com/github/codeql/blob/main/csharp/extractor/Semmle.Extraction.CIL/Entities/CustomAttributeDecoder.cs).

After the changes to make support for attributes on methods, I can confirm that generic attributes also works out of the box for attributes extracted from CIL.
To align a bit more with source code extracted attributes, I have changed the printing of constructed types extracted from CIL.

File: csharp/ql/lib/change-notes/2023-01-03-genericattributes.md (hunk 2 above)
Comment (michaelnebel) [reply csharp/ql/lib/change-notes/2023-01-03-genericattributes.md]: That is a good point! According to the [change note description](https://github.com/github/codeql/blob/main/docs/change-notes.md) for libraries, the category is probably `majorAnalysis`.

**Review by tamasvajk — APPROVED** [review]

**Comment by tamasvajk** [comment]
Let's check with DCA too. Otherwise it looks good to me.

**Comment by michaelnebel** [comment]
> Let's check with DCA too. Otherwise it looks good to me.

👍 Yes definitely! I will take this PR out of draft and rebase as soon as its dependency has been merged and then run DCA!

**Comment by michaelnebel** [comment]
DCA looks good!
@tamasvajk : Can you re-approve. The only changes since list time is a rebase and a merge of two commits containing the change-note.

**Review by tamasvajk — APPROVED** [review]

**Pull request merged**